File: backend/agent/tools/archive.py
from __future__ import annotations

import os
import tarfile
import zipfile
from typing import Any

from agent.tools.base import Tool, _in_cwd

_ARCHIVE_EXTS = {".zip", ".tar", ".tgz", ".tbz2", ".tar.gz", ".tar.bz2"}


def _detect_format(path: str) -> str | None:
    low = path.lower()
    if low.endswith(".zip"):
        return "zip"
    if low.endswith(".tar"):
        return "tar"
    if low.endswith(".tgz") or low.endswith(".tar.gz"):
        return "tar.gz"
    if low.endswith(".tbz2") or low.endswith(".tar.bz2"):
        return "tar.bz2"
    return None


def _tar_mode(fmt: str) -> str:
    return {
        "tar": "w:",
        "tar.gz": "w:gz",
        "tar.bz2": "w:bz2",
    }[fmt]


def _open_tar(path: str, fmt: str, mode: str = "r"):
    if mode == "r":
        m = {"tar": "r:", "tar.gz": "r:gz", "tar.bz2": "r:bz2"}[fmt]
    else:
        m = _tar_mode(fmt)
    return tarfile.open(path, m)


class ArchiveTool(Tool):
    """List, extract, or create zip/tar archives."""

    name = "archive"
    description = (
        "Work with zip and tar archives. "
        "action: 'list' (show contents), 'extract' (unpack to dest), 'create' (pack sources into path). "
        "Format is detected from the file extension. For 'create', the format is taken from the output extension "
        "(.zip, .tar, .tar.gz/.tgz, .tar.bz2/.tbz2). Extraction is blocked outside the workspace; creation requires sources inside it."
    )
    params = {
        "action": {"type": "string", "description": "list | extract | create"},
        "path": {"type": "string", "description": "Archive file path"},
        "dest": {
            "type": "string",
            "description": "Destination directory (for extract, default: archive's directory)",
        },
        "sources": {
            "type": "array",
            "description": "Files/dirs to pack (for create)",
            "items": {"type": "string"},
        },
    }

    def is_dangerous(self, args: dict[str, Any]) -> bool:
        # Both extract and create mutate the filesystem — confirm.
        # But list is read-only: only confirm if user is doing mutation.
        action = str(args.get("action", "")).lower()
        return action in ("extract", "create")

    async def execute(self, args: dict[str, Any]) -> str:
        action = str(args.get("action", "")).lower()
        path = args.get("path", "")
        if not path:
            return "error: path is required"
        if action not in ("list", "extract", "create"):
            return "error: action must be list, extract, or create"

        fmt = _detect_format(path)
        if fmt is None:
            exts = ", ".join(sorted(_ARCHIVE_EXTS))
            return f"error: cannot detect archive format from extension (supported: {exts})"

        try:
            if action == "list":
                return await self._list(path, fmt)
            if action == "extract":
                return await self._extract(path, fmt, args.get("dest", ""))
            return await self._create(path, fmt, args.get("sources", []))
        except Exception as e:
            return f"error: {e}"

    async def _list(self, path: str, fmt: str) -> str:
        if not os.path.exists(path):
            return f"error: not found: {path}"
        if not os.path.isfile(path):
            return f"error: not a file: {path}"

        if fmt == "zip":
            with zipfile.ZipFile(path) as zf:
                infos = zf.infolist()
            lines = []
            for info in infos[:200]:
                lines.append(f"{info.file_size:>10}  {info.date_time}  {info.filename}")
            if len(infos) > 200:
                lines.append(f"... and {len(infos) - 200} more")
            return "\n".join(lines) if lines else "(empty)"

        with _open_tar(path, fmt, "r") as tf:
            members = tf.getmembers()
        lines = []
        for m in members[:200]:
            lines.append(f"{m.size:>10}  {m.mtime:.0f}  {'d' if m.isdir() else 'f'}  {m.name}")
        if len(members) > 200:
            lines.append(f"... and {len(members) - 200} more")
        return "\n".join(lines) if lines else "(empty)"

    async def _extract(self, path: str, fmt: str, dest: str) -> str:
        # Refuse path traversal in members (zip slip / tar slip).
        # Both extract and create must land inside the workspace.
        dest = dest or os.path.dirname(os.path.abspath(path)) or "."
        if not _in_cwd(dest):
            return "error: dest is outside current directory"

        if fmt == "zip":
            with zipfile.ZipFile(path) as zf:
                # zip slip guard
                for info in zf.infolist():
                    target = os.path.realpath(os.path.join(dest, info.filename))
                    if not target.startswith(
                        os.path.realpath(dest) + os.sep
                    ) and target != os.path.realpath(dest):
                        return f"error: zip slip detected in member: {info.filename}"
                zf.extractall(dest)
            return f"extracted: {path} → {dest}"

        with _open_tar(path, fmt, "r") as tf:
            # tar slip guard
            for m in tf.getmembers():
                target = os.path.realpath(os.path.join(dest, m.name))
                if not target.startswith(
                    os.path.realpath(dest) + os.sep
                ) and target != os.path.realpath(dest):
                    return f"error: tar slip detected in member: {m.name}"
            tf.extractall(dest)
        return f"extracted: {path} → {dest}"

    async def _create(self, path: str, fmt: str, sources: list) -> str:
        if not sources:
            return "error: sources is required for create"
        # All sources must exist and live inside the workspace.
        missing = [s for s in sources if not os.path.exists(s)]
        if missing:
            return f"error: source(s) not found: {', '.join(missing)}"
        outside = [s for s in sources if not _in_cwd(s)]
        if outside:
            return f"error: source(s) outside current directory: {', '.join(outside)}"
        # Output must be inside the workspace too.
        if not _in_cwd(path):
            return "error: archive path is outside current directory"

        parent = os.path.dirname(os.path.abspath(path))
        if parent:
            os.makedirs(parent, exist_ok=True)

        count = 0
        if fmt == "zip":
            with zipfile.ZipFile(path, "w", zipfile.ZIP_DEFLATED) as zf:
                for src in sources:
                    if os.path.isdir(src):
                        for root, _, files in os.walk(src):
                            for f in files:
                                full = os.path.join(root, f)
                                arc = os.path.relpath(full, os.path.dirname(src) or ".").replace(
                                    "\\", "/"
                                )
                                zf.write(full, arc)
                                count += 1
                    else:
                        arc = os.path.basename(src)
                        zf.write(src, arc)
                        count += 1
            return f"created: {path} ({count} entries)"

        with _open_tar(path, fmt, "w") as tf:
            for src in sources:
                tf.add(
                    src,
                    arcname=os.path.basename(src) if not os.path.isdir(src) else None,
                    recursive=True,
                )
                count += 1
        return f"created: {path} ({count} entries)"

File: agent/tools/archive.go
package tools

import (
	"archive/tar"
	"archive/zip"
	"compress/bzip2"
	"compress/gzip"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
	"time"
)

type ArchiveTool struct{}

func (t *ArchiveTool) Name() string { return "archive" }

func (t *ArchiveTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "List, create, or extract archives (.zip, .tar, .tar.gz, .tar.bz2).",
		Params: map[string]any{
			"action":  prop("string", "One of: list, create, extract"),
			"path":    prop("string", "Archive file path"),
			"dest":    prop("string", "Destination directory (extract only)"),
			"sources": map[string]any{"type": "array", "items": map[string]any{"type": "string"}, "description": "Files/dirs to include (create only)"},
		},
		Required: []string{"action", "path"},
	}
}

func (t *ArchiveTool) Execute(args map[string]any) Result {
	action := strings.ToLower(getStr(args, "action"))
	path := getStr(args, "path")
	if path == "" {
		return R("error: path is required")
	}
	if action != "list" && action != "extract" && action != "create" {
		return R("error: action must be list, extract, or create")
	}

	format := detectFormat(path)
	if format == "" {
		return R("error: cannot detect archive format from extension (supported: .zip, .tar, .tgz, .tar.gz, .tbz2, .tar.bz2)")
	}

	switch action {
	case "list":
		return t.listArchive(path, format)
	case "extract":
		dest := getStr(args, "dest")
		return t.extractArchive(path, format, dest)
	case "create":
		sources, _ := args["sources"].([]any)
		return t.createArchive(path, format, sources)
	}
	return R("error: unknown action")
}

func (t *ArchiveTool) listArchive(path, format string) Result {
	if _, err := os.Stat(path); os.IsNotExist(err) {
		return R("error: not found: " + path)
	}
	fi, err := os.Stat(path)
	if err != nil {
		return R("error: " + err.Error())
	}
	if fi.IsDir() {
		return R("error: not a file: " + path)
	}

	if format == "zip" {
		r, err := zip.OpenReader(path)
		if err != nil {
			return R("error: " + err.Error())
		}
		defer r.Close()
		var b strings.Builder
		for i, f := range r.File {
			if i >= 200 {
				b.WriteString(fmt.Sprintf("\n... and %d more", len(r.File)-200))
				break
			}
			b.WriteString(fmt.Sprintf("%10d  %s  %s\n", f.UncompressedSize64, f.Modified.Format("2006-01-02 15:04:05"), f.Name))
		}
		if b.Len() == 0 {
			return R("(empty)")
		}
		return R(strings.TrimRight(b.String(), "\n"))
	}

	return t.listTar(path, format)
}

func (t *ArchiveTool) listTar(path, format string) Result {
	f, err := os.Open(path)
	if err != nil {
		return R("error: " + err.Error())
	}
	defer f.Close()

	var r io.Reader = f
	if format == "tar.gz" {
		gz, err := gzip.NewReader(f)
		if err != nil {
			return R("error: " + err.Error())
		}
		defer gz.Close()
		r = gz
	} else if format == "tar.bz2" {
		r = bzip2.NewReader(f)
	}

	tr := tar.NewReader(r)
	var b strings.Builder
	count := 0
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			return R("error: " + err.Error())
		}
		if count >= 200 {
			count++
			continue
		}
		kind := "f"
		if hdr.Typeflag == tar.TypeDir {
			kind = "d"
		}
		b.WriteString(fmt.Sprintf("%10d  %s  %s  %s\n", hdr.Size, time.Unix(hdr.ModTime.Unix(), 0).Format("2006-01-02 15:04:05"), kind, hdr.Name))
		count++
	}
	if count > 200 {
		b.WriteString(fmt.Sprintf("... and %d more", count-200))
	}
	if b.Len() == 0 {
		return R("(empty)")
	}
	return R(strings.TrimRight(b.String(), "\n"))
}

func (t *ArchiveTool) extractArchive(path, format, dest string) Result {
	if dest == "" {
		abs, err := filepath.Abs(path)
		if err != nil {
			return R("error: " + err.Error())
		}
		dest = filepath.Dir(abs)
	}
	if !inCwd(dest) {
		return R("error: dest is outside current directory")
	}

	if format == "zip" {
		r, err := zip.OpenReader(path)
		if err != nil {
			return R("error: " + err.Error())
		}
		defer r.Close()

		// zip slip guard
		destAbs, _ := filepath.Abs(dest)
		for _, f := range r.File {
			target := filepath.Join(destAbs, f.Name)
			if !strings.HasPrefix(filepath.Clean(target), destAbs+string(filepath.Separator)) && target != destAbs {
				return R("error: zip slip detected in member: " + f.Name)
			}
		}
		for _, f := range r.File {
			target := filepath.Join(destAbs, f.Name)
			if f.FileInfo().IsDir() {
				os.MkdirAll(target, 0755)
				continue
			}
			os.MkdirAll(filepath.Dir(target), 0755)
			src, err := f.Open()
			if err != nil {
				return R("error: " + err.Error())
			}
			dst, err := os.Create(target)
			if err != nil {
				src.Close()
				return R("error: " + err.Error())
			}
			io.Copy(dst, src)
			src.Close()
			dst.Close()
		}
		return R("extracted: " + path + " → " + dest)
	}

	return t.extractTar(path, format, dest)
}

func (t *ArchiveTool) extractTar(path, format, dest string) Result {
	f, err := os.Open(path)
	if err != nil {
		return R("error: " + err.Error())
	}
	defer f.Close()

	var r io.Reader = f
	if format == "tar.gz" {
		gz, err := gzip.NewReader(f)
		if err != nil {
			return R("error: " + err.Error())
		}
		defer gz.Close()
		r = gz
	} else if format == "tar.bz2" {
		r = bzip2.NewReader(f)
	}

	tr := tar.NewReader(r)
	destAbs, _ := filepath.Abs(dest)

	// tar slip guard: read all headers first
	var headers []*tar.Header
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			return R("error: " + err.Error())
		}
		target := filepath.Join(destAbs, hdr.Name)
		if !strings.HasPrefix(filepath.Clean(target), destAbs+string(filepath.Separator)) && target != destAbs {
			return R("error: tar slip detected in member: " + hdr.Name)
		}
		headers = append(headers, hdr)
	}

	// Re-open for extraction
	f.Seek(0, 0)
	if format == "tar.bz2" {
		// Need to re-open for compressed formats
		f.Close()
		f, err = os.Open(path)
		if err != nil {
			return R("error: " + err.Error())
		}
		defer f.Close()
		r = f
		if format == "tar.gz" {
			gz, _ := gzip.NewReader(f)
			defer gz.Close()
			r = gz
		} else {
			r = bzip2.NewReader(f)
		}
		tr = tar.NewReader(r)
	} else {
		f.Seek(0, 0)
		tr = tar.NewReader(f)
	}

	for _, hdr := range headers {
		target := filepath.Join(destAbs, hdr.Name)
		switch hdr.Typeflag {
		case tar.TypeDir:
			os.MkdirAll(target, 0755)
		case tar.TypeReg:
			os.MkdirAll(filepath.Dir(target), 0755)
			dst, err := os.Create(target)
			if err != nil {
				return R("error: " + err.Error())
			}
			if _, err := io.Copy(dst, tr); err != nil {
				dst.Close()
				return R("error: " + err.Error())
			}
			dst.Close()
		}
	}
	return R("extracted: " + path + " → " + dest)
}

func (t *ArchiveTool) createArchive(path, format string, sources []any) Result {
	if len(sources) == 0 {
		return R("error: sources is required for create")
	}

	var srcPaths []string
	var missing []string
	var outside []string
	for _, s := range sources {
		p := fmt.Sprint(s)
		if _, err := os.Stat(p); os.IsNotExist(err) {
			missing = append(missing, p)
		} else if !inCwd(p) {
			outside = append(outside, p)
		} else {
			srcPaths = append(srcPaths, p)
		}
	}
	if len(missing) > 0 {
		return R("error: source(s) not found: " + strings.Join(missing, ", "))
	}
	if len(outside) > 0 {
		return R("error: source(s) outside current directory: " + strings.Join(outside, ", "))
	}
	if !inCwd(path) {
		return R("error: archive path is outside current directory")
	}

	absPath, err := filepath.Abs(path)
	if err != nil {
		return R("error: " + err.Error())
	}
	if d := filepath.Dir(absPath); d != "" {
		os.MkdirAll(d, 0755)
	}

	if format == "zip" {
		return t.createZip(absPath, srcPaths)
	}
	return t.createTar(absPath, format, srcPaths)
}

func (t *ArchiveTool) createZip(path string, sources []string) Result {
	zf, err := os.Create(path)
	if err != nil {
		return R("error: " + err.Error())
	}
	defer zf.Close()

	zw := zip.NewWriter(zf)
	defer zw.Close()

	count := 0
	for _, src := range sources {
		fi, err := os.Stat(src)
		if err != nil {
			return R("error: " + err.Error())
		}
		if fi.IsDir() {
			filepath.Walk(src, func(p string, info os.FileInfo, err error) error {
				if err != nil || info.IsDir() {
					return nil
				}
				rel, _ := filepath.Rel(filepath.Dir(src), p)
				rel = strings.ReplaceAll(rel, "\\", "/")
				w, err := zw.Create(rel)
				if err != nil {
					return err
				}
				data, err := os.ReadFile(p)
				if err != nil {
					return err
				}
				w.Write(data)
				count++
				return nil
			})
		} else {
			w, err := zw.Create(filepath.Base(src))
			if err != nil {
				return R("error: " + err.Error())
			}
			data, err := os.ReadFile(src)
			if err != nil {
				return R("error: " + err.Error())
			}
			w.Write(data)
			count++
		}
	}
	return R(fmt.Sprintf("created: %s (%d entries)", path, count))
}

func (t *ArchiveTool) createTar(path, format string, sources []string) Result {
	f, err := os.Create(path)
	if err != nil {
		return R("error: " + err.Error())
	}
	defer f.Close()

	var w io.WriteCloser = f
	if format == "tar.gz" {
		gw := gzip.NewWriter(f)
		defer gw.Close()
		w = gw
	} else if format == "tar.bz2" {
		return R("error: bzip2 write not supported (use gzip instead)")
	}

	tw := tar.NewWriter(w)
	defer tw.Close()

	count := 0
	for _, src := range sources {
		fi, err := os.Stat(src)
		if err != nil {
			return R("error: " + err.Error())
		}
		if fi.IsDir() {
			filepath.Walk(src, func(p string, info os.FileInfo, err error) error {
				if err != nil {
					return nil
				}
				rel, _ := filepath.Rel(filepath.Dir(src), p)
				rel = strings.ReplaceAll(rel, "\\", "/")
				if info.IsDir() {
					hdr, _ := tar.FileInfoHeader(info, "")
					hdr.Name = rel + "/"
					tw.WriteHeader(hdr)
					return nil
				}
				hdr, _ := tar.FileInfoHeader(info, "")
				hdr.Name = rel
				if err := tw.WriteHeader(hdr); err != nil {
					return err
				}
				data, err := os.ReadFile(p)
				if err != nil {
					return err
				}
				tw.Write(data)
				count++
				return nil
			})
		} else {
			hdr, _ := tar.FileInfoHeader(fi, "")
			hdr.Name = filepath.Base(src)
			if err := tw.WriteHeader(hdr); err != nil {
				return R("error: " + err.Error())
			}
			data, err := os.ReadFile(src)
			if err != nil {
				return R("error: " + err.Error())
			}
			tw.Write(data)
			count++
		}
	}
	return R(fmt.Sprintf("created: %s (%d entries)", path, count))
}

var archiveExts = map[string]string{
	".zip":    "zip",
	".tar":    "tar",
	".tgz":    "tar.gz",
	".tar.gz": "tar.gz",
	".tbz2":   "tar.bz2",
	".tar.bz2": "tar.bz2",
}

func detectFormat(path string) string {
	low := strings.ToLower(path)
	for ext, fmt := range archiveExts {
		if strings.HasSuffix(low, ext) {
			return fmt
		}
	}
	return ""
}
